File: kattis/validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // min, max, isBST
    tuple<long,long,bool> dfs(TreeNode* root) {
        if (root == nullptr) return { 1e18, -1e18, true };
        
        long lMin, lMax, rMin, rMax;
        bool lisBST, risBST;
        
        tie(lMin, lMax, lisBST) = dfs(root->left);
        if (!lisBST) return { -1, -1, false };
        
        tie(rMin, rMax, risBST) = dfs(root->right);
        if (!risBST) return { -1, -1, false };
        
        long val = (long) root->val;
        bool isBST = (val > lMax) and (val < rMin);
        long minVal = min(val, min(lMin, rMin));
        long maxVal = max(val, max(lMax, rMax));
        
        return { minVal, maxVal, isBST };
    }
    
    bool isValidBST(TreeNode* root) {
        bool isBST;
        tie(ignore, ignore, isBST) = dfs(root);
        return isBST;
    }
};

File: kattis/validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        # iterate dfs inorder traversal
        stk = []
        last = None
        curr = root
        while stk or curr:
            if curr: # go left
                
                # preorder
                #print(curr.val)
                
                stk.append(curr)
                curr = curr.left
            else: # pop and go right
                curr = stk.pop()
                
                # inorder
                if last and last.val >= curr.val:
                    return False
                last = curr
                    
                curr = curr.right
            
        return True
"""
        5
    1       4
        3       6
"""